### PR TITLE
[Core] Implement sharded state loader

### PR DESCRIPTION
**Pull request by aurickq**
This PR implements a new model loader that directly loads the sharded states of each worker when using `DistributedGPUExecutor`. When using tensor parallelism, this avoids each worker reading the full checkpoint just to load a small shard of it. Our tests using Arctic (https://github.com/vllm-project/vllm/pull/4652) showed a 10x improvement in model loading speed from NVMe when using 8x tensor parallelism.

For quantization methods like DeepSpeed's FP_Quantize (also used in https://github.com/vllm-project/vllm/pull/4652) that quantize after loading, this PR allows easy creation of a quantized checkpoint that is directly loaded into each worker, further speeding up model loading.

This PR is separated out from https://github.com/vllm-project/vllm/pull/4652.

**BEFORE SUBMITTING, PLEASE READ THE CHECKLIST BELOW AND FILL IN THE DESCRIPTION ABOVE**

---

<details>

<summary><b> PR Checklist (Click to Expand) </b></summary>

<p>Thank you for your contribution to vLLM! Before submitting the pull request, please ensure the PR meets the following criteria. This helps vLLM maintain the code quality and improve the efficiency of the review process.</p>

<h3>PR Title and Classification</h3>
<p>Only specific types of PRs will be reviewed. The PR title is prefixed appropriately to indicate the type of change. Please use one of the following:</p>
<ul>
    <li><code>[Bugfix]</code> for bug fixes.</li>
    <li><code>[CI/Build]</code> for build or continuous integration improvements.</li>
    <li><code>[Doc]</code> for documentation fixes and improvements.</li>
    <li><code>[Model]</code> for adding a new model or improving an existing model. Model name should appear in the title.</li>
    <li><code>[Frontend]</code> For changes on the vLLM frontend (e.g., OpenAI API server, <code>LLM</code> class, etc.) </li>
    <li><code>[Kernel]</code> for changes affecting CUDA kernels or other compute kernels.</li>
    <li><code>[Core]</code> for changes in the core vLLM logic (e.g., <code>LLMEngine</code>, <code>AsyncLLMEngine</code>, <code>Scheduler</code>, etc.)</li>
    <li><code>[Hardware][Vendor]</code> for hardware-specific changes. Vendor name should appear in the prefix (e.g., <code>[Hardware][AMD]</code>).</li>
    <li><code>[Misc]</code> for PRs that do not fit the above categories. Please use this sparingly.</li>
</ul>
<p><strong>Note:</strong> If the PR spans more than one category, please include all relevant prefixes.</p>

<h3>Code Quality</h3>

<p>The PR need to meet the following code quality standards:</p>

<ul>
    <li>We adhere to <a href="https://google.github.io/styleguide/pyguide.html">Google Python style guide</a> and <a href="https://google.github.io/styleguide/cppguide.html">Google C++ style guide</a>.</li>
    <li>Pass all linter checks. Please use <a href="https://github.com/vllm-project/vllm/blob/main/format.sh"><code>format.sh</code></a> to format your code.</li>
    <li>The code need to be well-documented to ensure future contributors can easily understand the code.</li>
    <li>Include sufficient tests to ensure the project to stay correct and robust. This includes both unit tests and integration tests.</li>
    <li>Please add documentation to <code>docs/source/</code> if the PR modifies the user-facing behaviors of vLLM. It helps vLLM user understand and utilize the new features or changes.</li>
</ul>

<h3>Notes for Large Changes</h3>
<p>Please keep the changes as concise as possible. For major architectural changes (>500 LOC excluding kernel/data/config/test), we would expect a GitHub issue (RFC) discussing the technical design and justification. Otherwise, we will tag it with <code>rfc-required</code> and might not go through the PR.</p>

<h3>What to Expect for the Reviews</h3>

<p>The goal of the vLLM team is to be a <i>transparent reviewing machine</i>. We would like to make the review process transparent and efficient and make sure no contributor feel confused or frustrated. However, the vLLM team is small, so we need to prioritize some PRs over others. Here is what you can expect from the review process: </p>

<ul>
    <li> After the PR is submitted, the PR will be assigned to a reviewer. Every reviewer will pick up the PRs based on their expertise and availability.</li>
    <li> After the PR is assigned, the reviewer will provide status update every 2-3 days. If the PR is not reviewed within 7 days, please feel free to ping the reviewer or the vLLM team.</li>
    <li> After the review, the reviewer will put an <code> action-required</code> label on the PR if there are changes required. The contributor should address the comments and ping the reviewer to re-review the PR.</li>
    <li> Please respond to all comments within a reasonable time frame. If a comment isn't clear or you disagree with a suggestion, feel free to ask for clarification or discuss the suggestion.
 </li>
</ul>

<h3>Thank You</h3>

<p> Finally, thank you for taking the time to read these guidelines and for your interest in contributing to vLLM. Your contributions make vLLM a great tool for everyone! </p>

</details>

### CODE REVIEW

**Comment by Yard1** [comment]
We should add a test to ensure it's working correctly (loaded weights and outputs are the same as with the default loader)

File: examples/save_sharded_state.py [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+    # Dump worker states to output directory
+    model_executor = llm.llm_engine.model_executor
+    model_executor.save_sharded_state(path=args.output,
+                                      pattern=args.pattern,
+                                      max_size=5 * 1024**3)
+    # Copy metadata files to output directory
+    for file in os.listdir(model_path):
+        if not any(
+                file.endswith(ext) for ext in (".bin", ".pt", ".safetensors")):
+            shutil.copy(f"{model_path}/{file}", args.output)
```
Comment (mgoin) [inline examples/save_sharded_state.py:62]: shouldn't this set the config.json or quant_config.json next to the model weights to inform vLLM loading the model what type of quantization the model checkpoint is in?

**Comment by aurickq** [comment]
> We should add a test to ensure it's working correctly (loaded weights and outputs are the same as with the default loader)

Sure, that makes sense. What's the spec of the test runner machine we should target? OK to assume cuda device is present?

File: examples/save_sharded_state.py (hunk 1 above)
Comment (aurickq) [reply examples/save_sharded_state.py:62]: Currently, it's just copying the config.json and quant_config.json from the input checkpoint that's being converted, which works for the use cases we've tested. Actually, I am not sure if it's correct to override these configs (or add a quant_config.json where it wasn't there previously) because then the config may mismatch the final loaded states?

**Comment by Yard1** [comment]
@aurickq IIRC we are using L4 GPUs in CI, @simon-mo to confirm

**Comment by sfc-gh-aqiao** [comment]
@Yard1 added test, please take a look

**Comment by aurickq** [comment]
@Yard1 added test, please take a look

File: examples/save_sharded_state.py [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+        raise ValueError("model path must be a local directory")
+    # Create LLM instance from arguments
+    llm = LLM(**dataclasses.asdict(engine_args))
+    # Prepare output directory
+    Path(args.output).mkdir(exist_ok=True)
+    # Dump worker states to output directory
+    model_executor = llm.llm_engine.model_executor
+    model_executor.save_sharded_state(path=args.output,
+                                      pattern=args.pattern,
+                                      max_size=5 * 1024**3)
```
Comment (Yard1) [inline examples/save_sharded_state.py:57]: make max size configurable with this as the default?

File: examples/save_sharded_state.py [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+    # Copy metadata files to output directory
+    for file in os.listdir(model_path):
+        if not any(
+                file.endswith(ext) for ext in (".bin", ".pt", ".safetensors")):
```
Comment (Yard1) [inline examples/save_sharded_state.py:61]: use `os.path.splitext` instead

File: tests/test_sharded_state_loader.py [hunk 4]
```diff
@@ -0,0 +1,55 @@
...
+]
+
+# Create a sampling params object.
+sampling_params = SamplingParams(temperature=0.8, top_p=0.95, seed=0)
+
+
+def test_sharded_state_loader():
+    weights_patterns = ("*.bin", "*.pt", "*.safetensors")
+
+    cache_dir = TemporaryDirectory()
```
Comment (Yard1) [inline tests/test_sharded_state_loader.py:23]: use `TemporaryDirectory` as context manager

File: tests/test_sharded_state_loader.py [hunk 5]
```diff
@@ -0,0 +1,55 @@
...
+
+prompts = [
+    "Hello, my name is",
+    "The president of the United States is",
+    "The capital of France is",
+    "The future of AI is",
+]
+
+# Create a sampling params object.
+sampling_params = SamplingParams(temperature=0.8, top_p=0.95, seed=0)
```
Comment (Yard1) [inline tests/test_sharded_state_loader.py:17]: increase max tokens to something large and ignore eos

File: vllm/model_executor/model_loader/loader.py [hunk 6]
```diff
@@ -347,6 +347,105 @@ def load_model(self, *, model_config: ModelConfig,
...
+                # TODO: support un-sharded checkpoints too
+                raise ValueError(
+                    f"Could not find checkpoint files '{pattern}', only "
+                    f"pre-sharded checkpoints are currently supported!"
+                )
+            state_dict = dict(model.state_dict())
+            data_ptrs = {}
+            for path in filepaths:
+                for key, val in load_file(path).items():
+                    data_ptrs[state_dict[key].data_ptr()] = key
```
Comment (Yard1) [inline vllm/model_executor/model_loader/loader.py:399]: hmm is this safe? what if we have an incomplete view or something like that? eg. tensor A has shape [200] and tensor B has shape [500], with both starting at the same data pointer. then we will skip copying tensor B, which is unwanted.

**Comment by simon-mo** [comment]
One question I have is that can this be implemented using safetensor's partial read? safetensors have all the metadata in headers so you can access the tensors partially

File: examples/save_sharded_state.py (hunk 2 above)
Comment (aurickq) [reply examples/save_sharded_state.py]: done

File: vllm/model_executor/model_loader/loader.py (hunk 6 above)
Comment (aurickq) [reply vllm/model_executor/model_loader/loader.py]: nice catch! fixed it

**Comment by aurickq** [comment]
> One question I have is that can this be implemented using safetensor's partial read? safetensors have all the metadata in headers so you can access the tensors partially

Conceptually, I think so. Though for larger models like Arctic we prefer this implementation for a few reasons:

1. With the current implementation and multi-node inference, we may download only the required files to each node.
2. The deepspeedfp quantizer will reshape the parameters and may change its number of elements, which makes it a bit tricky to calculate the right slice to load (need to understand internals of the quantizer to slice the quantized weights).

**Comment by njhill** [comment]
@aurickq curious how this relates to https://github.com/vllm-project/vllm/pull/3729?

**Comment by aurickq** [comment]
> @aurickq curious how this relates to #3729?

Ah, I hadn't seen that PR, thanks for bringing it up! From what I understand, both PRs address the problem of model loading speed, particularly for larger models, but the approaches seem pretty different. #3729 modifies the existing model loading path so each worker loads a different set of tensors, then broadcasts the shards to each other. This PR loads the model once using the default path and dumps each worker state to disk, then subsequent model loads can direct read these states, skipping the default model loading path altogether.

Some first thoughts on the pros/cons:

- This PR requires the creation of a new checkpoint (the worker states), and this checkpoint is only compatible with a fixed configuration (parallelism, quantization, etc.). Whereas #3729 seems more flexible in this regard.
- #3729 seems to require a lot more changes in vllm and in each model definition, whereas this PR only requires a handful of lines of change in vllm, and no changes in any of the models (so any future model is automatically supported).
- #3729 seems to be more coupled with the parallelism implementation (e.g. tensor parallel in this case). However we are currently implementing pipeline-parallelism to support multi-node inference, which also will be automatically supported by this PR.

File: vllm/model_executor/model_loader/loader.py [hunk 7]
```diff
@@ -347,6 +347,124 @@ def load_model(self, *, model_config: ModelConfig,
...
+        extra_config = ({} if load_config.model_loader_extra_config is None
+                        else load_config.model_loader_extra_config.copy())
+        self.pattern = extra_config.pop("pattern", self.DEFAULT_PATTERN)
+        if extra_config:
+            raise ValueError(f"Unexpected extra config keys for load format "
+                             f"{load_config.load_format}: "
+                             f"{load_config.model_loader_extra_config.keys()}")
+
+    @staticmethod
+    def _filter_subtensors(tensors: Dict[str, torch.Tensor]):
```
Comment (Yard1) [inline vllm/model_executor/model_loader/loader.py:371]: can we unit test this method?

File: vllm/model_executor/model_loader/loader.py [hunk 8]
```diff
@@ -347,6 +347,124 @@ def load_model(self, *, model_config: ModelConfig,
...
+        """
+        Filter out all tensors that share the same memory or a subset of the
+        memory of another tensor.
+        """
+        from safetensors.torch import storage_ptr, storage_size
+        result = {}
+        for key1, tensor1 in tensors.items():
```
Comment (Yard1) [inline vllm/model_executor/model_loader/loader.py:378]: this looks suspect from a performance perspective (O(n^2)) but I guess it should be fine...

File: vllm/model_executor/model_loader/loader.py [hunk 9]
```diff
@@ -347,6 +347,124 @@ def load_model(self, *, model_config: ModelConfig,
...
+                # TODO: support un-sharded checkpoints too
+                raise ValueError(
+                    f"Could not find checkpoint files '{pattern}', only "
+                    f"pre-sharded checkpoints are currently supported!"
+                )
+            state_dict = self._filter_subtensors(model.state_dict())
+            for path in filepaths:
+                for key, tensor in load_file(path).items():
+                    state_dict[key].copy_(tensor)
+                    state_dict.pop(key)
```
Comment (Yard1) [inline vllm/model_executor/model_loader/loader.py:424]: this will load the entire file eagerly, better approach should be:
```suggestion
            for path in filepaths:
                with safe_open(filename, framework="pt", device=device) as f:
                    for key in state_dict:
                        state_dict[key].copy_(f.get_tensor(key))
```

**Review by Yard1 — COMMENTED** [review]
Thanks, two last comments

File: vllm/model_executor/model_loader/loader.py (hunk 8 above)
Comment (zhisbug) [reply vllm/model_executor/model_loader/loader.py]: this should be fine -- as a model at most just has 100 tensors..

Comment (aurickq) [reply vllm/model_executor/model_loader/loader.py]: Updated this part as there were some other issues. Now we only do the n^2 loop over tensors that share the same underlying storage, which should be much fewer than before.

File: vllm/model_executor/model_loader/loader.py [hunk 10]
```diff
@@ -347,6 +348,136 @@ def load_model(self, *, model_config: ModelConfig,
...
+                # TODO: support un-sharded checkpoints too
+                raise ValueError(
+                    f"Could not find checkpoint files '{pattern}', only "
+                    f"pre-sharded checkpoints are currently supported!"
+                )
+            state_dict = self._filter_subtensors(model.state_dict())
+            for path in filepaths:
+                with safe_open(path, framework="pt") as f:
+                    for key in f.keys():  # noqa: SIM118
+                        state_dict[key].copy_(f.get_tensor(key))
```
Comment (sfc-gh-zhwang) [inline vllm/model_executor/model_loader/loader.py:436]:
```
state_dict[key][:f.get_tensor(key).shape[0]].copy_(f.get_tensor(key))
```
I think you need to this to make it work for Lora, similar to https://github.com/vllm-project/vllm/blob/4e12131089f192334f6e09c8fe5cd85af1e25327/vllm/model_executor/layers/vocab_parallel_embedding.py#L87

Comment (sfc-gh-zhwang) [inline vllm/model_executor/model_loader/loader.py:436]:
```
state_dict[key][:f.get_tensor(key).shape[0]].copy_(f.get_tensor(key))
```
I think you need to do this to be compatible when `enable_lora` is true.

Comment (sfc-gh-zhwang) [reply vllm/model_executor/model_loader/loader.py:436]: similar to https://github.com/vllm-project/vllm/blob/e254497b66dcd87038969b0ad34d34425edfc5fe/vllm/model_executor/layers/vocab_parallel_embedding.py#L87

Comment (aurickq) [reply vllm/model_executor/model_loader/loader.py:436]: thanks for catching this! yeah there seems to be some issues with this pr in supporting lora. basically, during initialization, two things happen in vllm in this order:

1. The model loader is called to load the base model weights.
2. Lora is initialized, which may rename some parameters and reshape some weights.

currently, the worker states are saved after step (2). then, when later loading it through the model loader in step (1), it will contain some unexpected parameter names and shapes.

instead, the sharded weights should be both saved and loaded between (1) and (2), or both after step (2). the former might look something like

```python
llm = LLM(model_path, save_sharded_states=path_to_shards, ...)
```

then when loading

```python
llm = LLM(model_path, load_format="sharded_state",
          model_loader_extra_config=ShardedStateConfig(path_to_shards), ...)
```

and the latter might look like:

```python
llm = LLM(model_path, ...)

llm.save_sharded_states(path_to_shards)
```

then when loading

```python
llm = LLM(model_path, load_format="dummy", ...)

llm.load_sharded_states(path_to_shards)
```

anyone have any thoughts?

Comment (Yard1) [reply vllm/model_executor/model_loader/loader.py:436]: I think the shape trick should be fine. Any post-initialization modifications can only extend existing tensors or add new ones. There should not be a situation where the shape or information stored in a tensor is different. I think we will be fine with just a warning (and comments) in loading code when that happens.

Comment (aurickq) [reply vllm/model_executor/model_loader/loader.py]: done. btw, any small model that supports lora that can be used for test? lors is not supported for opt.

Comment (Yard1) [reply vllm/model_executor/model_loader/loader.py]: we are using llama 2 7b with lora for testing. it's not the smallest, but that's what we got right now :(

File: examples/save_sharded_state.py [hunk 11]
```diff
@@ -0,0 +1,76 @@
...
+    model_executor = llm.llm_engine.model_executor
+    model_executor.save_sharded_state(path=args.output,
+                                      pattern=args.file_pattern,
+                                      max_size=args.max_file_size)
+    # Copy metadata files to output directory
+    for file in os.listdir(model_path):
+        if os.path.splitext(file)[1] not in (".bin", ".pt", ".safetensors"):
+            if os.path.isdir(f"{model_path}/{file}"):
+                shutil.copytree(f"{model_path}/{file}",
+                                f"{args.output}/{file}")
```
Comment (Yard1) [inline examples/save_sharded_state.py:69]: suggestion: use os.path.join

File: vllm/model_executor/model_loader/loader.py (hunk 10 above)
Comment (aurickq) [reply vllm/model_executor/model_loader/loader.py]: should be all good now. thanks for the comments!

**Review by WoosukKwon — APPROVED** [review]
LGTM! The code looks clean and mergable. Thanks for addressing the comments from @Yard1.
